Foundation/FileManager+Win32.swift

```diff
@@ -689,27 +690,52 @@ extension FileManager {
                 } while FindNextFileW(h, &ffd)
                 return files
             }
-            while let url = _stack.popLast() {
-                if url.hasDirectoryPath && !_options.contains(.skipsSubdirectoryDescendants) {
-                    guard let dirContents = contentsOfDir(directory: url)?.reversed() else {
-                        if let handler = _errorHandler {
-                           let dirFSR = url.withUnsafeFileSystemRepresentation { $0.flatMap { fsr in String(utf8String: fsr) } }
-                           let keepGoing = handler(URL(fileURLWithPath: dirFSR ?? ""),
-                               _NSErrorWithWindowsError(GetLastError(), reading: true))
-                           if !keepGoing { return nil }
-                        }
-                        continue
-                    }
-                    _stack.append(contentsOf: dirContents)
+
+            func findFirstValidFile() -> URL? {
+              while let url = _stack.popLast() {
+                let dirFSR = url.withUnsafeFileSystemRepresentation { $0.flatMap { fsr in String(utf8String: fsr) } } ?? ""
+                if !FileManager.default.fileExists(atPath: dirFSR, isDirectory: nil) {
+                  if let handler = _errorHandler {
+                    let keepGoing = handler(URL(fileURLWithPath: dirFSR),
+                                    _NSErrorWithWindowsError(GetLastError(), reading: true))
+                    if !keepGoing { return nil }
+                  }
+                  continue
                 }
+
                 _current = url
-                return url
+                return _current
+              }
+              return nil
             }
-            return nil
+
+            var isDir: ObjCBool = false
+            let dirFSR = _current.withUnsafeFileSystemRepresentation { $0.flatMap { fsr in String(utf8String: fsr) } } ?? ""
```

Comment:
Again, the `dirFSR` is really misleading here.  Can we rename this as we do not know what the path points to.